Firestore/core/src/firebase/firestore/local/serializer.cc

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/local/serializer.h"
+#include "Firestore/Protos/nanopb/firestore/local/maybe_document.nanopb.h"
+#include "Firestore/Protos/nanopb/google/firestore/v1beta1/document.nanopb.h"
+#include "Firestore/core/src/firebase/firestore/model/no_document.h"
+#include "Firestore/core/src/firebase/firestore/nanopb/tag.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using firebase::firestore::nanopb::Reader;
+using firebase::firestore::nanopb::Tag;
+using firebase::firestore::nanopb::Writer;
+using firebase::firestore::util::Status;
+
+Status Serializer::EncodeMaybeDocument(const model::MaybeDocument& document,
+                                       std::vector<uint8_t>* out_bytes) const {
+  Writer writer = Writer::Wrap(out_bytes);
+  EncodeMaybeDocument(&writer, document);
+  return writer.status();
+}
+
+void Serializer::EncodeMaybeDocument(
+    Writer* writer, const model::MaybeDocument& maybe_doc) const {
+  switch (maybe_doc.type()) {
+    case model::MaybeDocument::Type::Document:
+      writer->WriteTag(
+          {PB_WT_STRING, firestore_client_MaybeDocument_document_tag});
+      writer->WriteNestedMessage([&](Writer* writer) {
+        const model::Document& doc =
```

Comment:
Optional: I think that when casting, it's better to use `auto`:
* type is clear from the statement;
* avoids duplication;
* eliminates potential for double-conversion:
```cpp
double x{};
int y = static_cast<float>(x);
```
Feel free to leave as is, though.